File: mobile/ios/Shared/RoutingConfiguration.swift
import Foundation

/// How much of the device's traffic the tunnel carries.
///
/// The mode is deliberately coarse. Which destinations are kept off the tunnel
/// is not a mode — it is a set of rules, and those live in
/// `RoutingConfiguration` so that turning the tunnel global does not mean
/// dismantling a bypass list the user spent time building.
enum RoutingMode: String, Codable, CaseIterable, Identifiable, Sendable {
    /// Everything goes through the tunnel. Bypass rules stay stored but idle.
    case allTraffic = "all-traffic"
    /// Everything except the enabled bypass rules goes through the tunnel.
    case bypassRules = "bypass-rules"

    var id: String { rawValue }

    var title: String {
        switch self {
        case .allTraffic:
            return "Route all traffic"
        case .bypassRules:
            return "Use bypass rules"
        }
    }

    var detail: String {
        switch self {
        case .allTraffic:
            return "Every IPv4, IPv6, and DNS destination goes through the selected " +
                "Queqiao provider. Bypass rules below are kept but not applied."
        case .bypassRules:
            return "Everything goes through Queqiao except the destinations matched " +
                "by the rules below."
        }
    }
}

/// Everything that decides where one profile's traffic goes.
///
/// This used to be three settings that did not know about each other: a
/// two-value traffic policy, a bundled-set toggle, and a typed route list. Two
/// of them applied regardless of what the third said, so "All traffic" was not
/// true whenever a bypass route or the country set was also on. They are one
/// type now because they were always answering one question.
///
/// Queqiao routes on addresses only. DNS resolves through the tunnel, so no
/// rule here can match a domain name.
struct RoutingConfiguration: Equatable, Sendable {
    var mode: RoutingMode = .allTraffic
    /// Private and link-local destinations, as one rule rather than as a mode.
    var bypassLocalNetworks = false
    /// The bundled registry set for China. Experimental and approximate: see
    /// CountryRoutes.
    var bypassChinaDirect = false
    /// Destinations typed by the user, in canonical CIDR text.
    var customRoutes: [String] = []
    /// The rule list as the user wrote or imported it, one rule per line. Kept
    /// here rather than beside it because a rule list is also an answer to
    /// "where does this profile's traffic go", and the whole point of this type
    /// is that there is one answer. The core parses and acts on the text; the
    /// route plan below ignores it.
    var rules: String = ""

    /// Whether the rules below the mode picker currently affect anything. The
    /// UI keeps them editable either way, and says plainly when they are idle.
    var rulesApply: Bool { mode == .bypassRules }

    /// Whether any rule is switched on, regardless of the mode. Used to tell
    /// "no rules configured" apart from "rules configured but not applied".
    var hasEnabledRules: Bool {
        bypassLocalNetworks || bypassChinaDirect || !customRoutes.isEmpty
    }

    /// How many lines the rule list holds, ignoring blanks and comments, for
    /// screens that report on it without parsing it.
    var ruleLineCount: Int {
        rules.split(whereSeparator: \.isNewline)
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty && !$0.hasPrefix("#") }
            .count
    }

    /// A one-line description for the connection summary card. Names the rules
    /// actually in force rather than the mode alone, because "use bypass rules"
    /// with every rule off carries exactly as much traffic as routing all of it
    /// and should not read differently.
    var summary: String {
        var parts: [String] = []
        if ruleLineCount > 0 {
            parts.append("\(ruleLineCount) \(ruleLineCount == 1 ? "rule" : "rules")")
        }
        guard rulesApply, hasEnabledRules else {
            return parts.isEmpty
                ? RoutingMode.allTraffic.title
                : "Routing by " + parts.formatted(.list(type: .and))
        }
        if bypassLocalNetworks { parts.append("local networks") }
        if bypassChinaDirect { parts.append("Chinese addresses") }
        if !customRoutes.isEmpty {
            parts.append("\(customRoutes.count) custom \(customRoutes.count == 1 ? "route" : "routes")")
        }
        return "Bypassing " + parts.formatted(.list(type: .and))
    }
}

extension RoutePlan {
    /// The plan a routing configuration installs.
    ///
    /// The country set is passed in rather than read here because loading it
    /// can fail, and the packet tunnel and the settings screen want to report
    /// that failure differently. Both still compose the plan through this one
    /// function, so what the screen previews is what the tunnel installs.
    static func make(
        for routing: RoutingConfiguration,
        chinaDirect: [IPPrefix] = [],
        limit: Int = defaultLimit
    ) -> RoutePlan {
        guard routing.rulesApply else {
            return make(userRoutes: [], builtIn: [], limit: limit)
        }
        var userRoutes = routing.customRoutes
        if routing.bypassLocalNetworks {
            userRoutes = localNetworks + userRoutes
        }
        return make(
            userRoutes: userRoutes,
            builtIn: routing.bypassChinaDirect ? chinaDirect : [],
            limit: limit
        )
    }
}

File: mobile/android/app/src/main/java/io/github/bojieli/queqiao/RoutingConfiguration.java
package io.github.bojieli.queqiao;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * One profile's routing settings. The same model as RoutingConfiguration.swift:
 * a mode that says whether address bypasses apply, the bypasses themselves, and
 * a rule list the core evaluates whichever mode is chosen.
 */
final class RoutingConfiguration {
    /** How many hand-entered bypass routes one profile may hold; iOS uses the same cap. */
    static final int MAX_CUSTOM_ROUTES = 256;

    enum Mode {
        ALL_TRAFFIC(
                "all-traffic",
                "Route all traffic",
                "Every IPv4, IPv6, and DNS destination goes through the selected Queqiao provider. "
                        + "Bypass rules below are kept but not applied."),
        BYPASS_RULES(
                "bypass-rules",
                "Use bypass rules",
                "Everything goes through Queqiao except the destinations matched by the rules below.");

        final String wireValue;
        final String title;
        final String detail;

        Mode(String wireValue, String title, String detail) {
            this.wireValue = wireValue;
            this.title = title;
            this.detail = detail;
        }

        static Mode fromWireValue(String value) {
            for (Mode mode : values()) {
                if (mode.wireValue.equals(value)) {
                    return mode;
                }
            }
            return ALL_TRAFFIC;
        }
    }

    static final RoutingConfiguration DEFAULT =
            new RoutingConfiguration(Mode.ALL_TRAFFIC, false, false, Collections.emptyList(), "");

    // What a build from before routing rules stored; still written so that build can read the catalog.
    private static final String LEGACY_ALL_TRAFFIC = "all-traffic";
    private static final String LEGACY_EXCLUDE_LOCAL = "exclude-local-networks";

    final Mode mode;
    final boolean bypassLocalNetworks;
    final boolean bypassChinaDirect;
    final List<String> customRoutes;
    final String rules;

    RoutingConfiguration(
            Mode mode,
            boolean bypassLocalNetworks,
            boolean bypassChinaDirect,
            List<String> customRoutes,
            String rules) {
        this.mode = mode;
        this.bypassLocalNetworks = bypassLocalNetworks;
        this.bypassChinaDirect = bypassChinaDirect;
        this.customRoutes = Collections.unmodifiableList(new ArrayList<>(customRoutes));
        this.rules = rules;
    }

    RoutingConfiguration withMode(Mode replacement) {
        return new RoutingConfiguration(replacement, bypassLocalNetworks, bypassChinaDirect, customRoutes, rules);
    }

    RoutingConfiguration withBypassLocalNetworks(boolean enabled) {
        return new RoutingConfiguration(mode, enabled, bypassChinaDirect, customRoutes, rules);
    }

    RoutingConfiguration withBypassChinaDirect(boolean enabled) {
        return new RoutingConfiguration(mode, bypassLocalNetworks, enabled, customRoutes, rules);
    }

    RoutingConfiguration withCustomRoutes(List<String> routes) {
        return new RoutingConfiguration(mode, bypassLocalNetworks, bypassChinaDirect, routes, rules);
    }

    RoutingConfiguration withRules(String text) {
        return new RoutingConfiguration(mode, bypassLocalNetworks, bypassChinaDirect, customRoutes, text);
    }

    boolean rulesApply() {
        return mode == Mode.BYPASS_RULES;
    }

    boolean hasEnabledRules() {
        return bypassLocalNetworks || bypassChinaDirect || !customRoutes.isEmpty();
    }

    int ruleLineCount() {
        int count = 0;
        for (String line : rules.split("\\R")) {
            String trimmed = line.trim();
            if (!trimmed.isEmpty() && !trimmed.startsWith("#")) {
                count++;
            }
        }
        return count;
    }

    String summary() {
        List<String> parts = new ArrayList<>();
        int ruleLines = ruleLineCount();
        if (ruleLines > 0) {
            parts.add(ruleLines + (ruleLines == 1 ? " rule" : " rules"));
        }
        if (!rulesApply() || !hasEnabledRules()) {
            return parts.isEmpty() ? Mode.ALL_TRAFFIC.title : "Routing by " + joinList(parts);
        }
        if (bypassLocalNetworks) {
            parts.add("local networks");
        }
        if (bypassChinaDirect) {
            parts.add("Chinese addresses");
        }
        if (!customRoutes.isEmpty()) {
            parts.add(customRoutes.size() + (customRoutes.size() == 1 ? " custom route" : " custom routes"));
        }
        return "Bypassing " + joinList(parts);
    }

    private static String joinList(List<String> parts) {
        if (parts.size() == 1) {
            return parts.get(0);
        }
        if (parts.size() == 2) {
            return parts.get(0) + " and " + parts.get(1);
        }
        return String.join(", ", parts.subList(0, parts.size() - 1)) + ", and " + parts.get(parts.size() - 1);
    }

    void writeTo(JSONObject record) throws JSONException {
        record.put("traffic_policy",
                rulesApply() && bypassLocalNetworks ? LEGACY_EXCLUDE_LOCAL : LEGACY_ALL_TRAFFIC);
        record.put("routing_mode", mode.wireValue);
        record.put("bypass_local_networks", bypassLocalNetworks);
        record.put("bypass_china_direct", bypassChinaDirect);
        record.put("bypass_routes", new JSONArray(customRoutes));
        record.put("routing_rules", rules);
    }

    static RoutingConfiguration readFrom(JSONObject record) {
        boolean legacyExcludesLocal = LEGACY_EXCLUDE_LOCAL.equals(record.optString("traffic_policy"));
        if (!record.has("routing_mode")) {
            // The old policy only ever decided whether local networks were excluded, so a
            // catalog that excluded them has to load as bypass rules or upgrading would
            // push those destinations back through the tunnel.
            return new RoutingConfiguration(
                    legacyExcludesLocal ? Mode.BYPASS_RULES : Mode.ALL_TRAFFIC,
                    legacyExcludesLocal, false, Collections.emptyList(), "");
        }
        List<String> routes = new ArrayList<>();
        JSONArray stored = record.optJSONArray("bypass_routes");
        for (int index = 0; stored != null && index < stored.length(); index++) {
            routes.add(stored.optString(index));
        }
        return new RoutingConfiguration(
                Mode.fromWireValue(record.optString("routing_mode")),
                record.optBoolean("bypass_local_networks"),
                record.optBoolean("bypass_china_direct"),
                routes,
                record.optString("routing_rules"));
    }
}
